File: MiApp_Android/app/src/main/java/com/miproyecto/viewmodel/SessionViewModel.kt
package com.miproyecto.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.miproyecto.model.ApiClient
import com.miproyecto.model.AuthTokens
import com.miproyecto.model.SessionStore
import com.miproyecto.model.User
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class SessionViewModel(application: Application) : AndroidViewModel(application) {
    private val store = SessionStore(application)

    private val _isAuthenticated = MutableStateFlow(false)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated

    private val _user = MutableStateFlow<User?>(null)
    val user: StateFlow<User?> = _user

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error

    private val _loading = MutableStateFlow(false)
    val loading: StateFlow<Boolean> = _loading

    fun login(email: String, password: String) {
        viewModelScope.launch {
            _loading.value = true
            _error.value = null
            try {
                val tokens = ApiClient.service.login(mapOf("email" to email, "password" to password))
                persist(tokens)
                _user.value = User(id = "local", name = email.substringBefore("@"), email = email)
                _isAuthenticated.value = true
            } catch (ex: Exception) {
                _error.value = "No se pudo iniciar sesión."
            }
            _loading.value = false
        }
    }

    fun signup(name: String, email: String, password: String) {
        viewModelScope.launch {
            _loading.value = true
            _error.value = null
            try {
                val tokens = ApiClient.service.signup(
                    mapOf("name" to name, "email" to email, "password" to password)
                )
                persist(tokens)
                _user.value = User(id = "local", name = name, email = email)
                _isAuthenticated.value = true
            } catch (ex: Exception) {
                _error.value = "No se pudo registrar la cuenta."
            }
            _loading.value = false
        }
    }

    fun logout() {
        viewModelScope.launch {
            store.clear()
            _isAuthenticated.value = false
            _user.value = null
        }
    }

    fun loadPersistedSession() {
        viewModelScope.launch {
            val token = store.readAccessToken()
            _isAuthenticated.value = token != null
        }
    }

    suspend fun readAccessToken(): String? {
        return store.readAccessToken()
    }

    private suspend fun persist(tokens: AuthTokens) {
        store.save(tokens)
    }
}

File: MiApp_iOS/App/ViewModels/SessionViewModel.swift
import Foundation
import SwiftUI

@MainActor
final class SessionViewModel: ObservableObject {
    @Published var isAuthenticated = false
    @Published var user: User?
    @Published var authError: String?
    @Published var isLoading = false

    private let apiClient = APIClient()
    private let keychain = KeychainStore()
    private let tokenService = "com.norteneo.auth"

    func login(email: String, password: String) async {
        isLoading = true
        authError = nil
        do {
            let tokens = try await apiClient.login(email: email, password: password)
            try persist(tokens: tokens)
            isAuthenticated = true
            user = User(id: UUID(), name: email.components(separatedBy: "@").first ?? "Usuario", email: email)
        } catch {
            authError = "No se pudo iniciar sesión."
        }
        isLoading = false
    }

    func signup(name: String, email: String, password: String) async {
        isLoading = true
        authError = nil
        do {
            let tokens = try await apiClient.signup(name: name, email: email, password: password)
            try persist(tokens: tokens)
            isAuthenticated = true
            user = User(id: UUID(), name: name, email: email)
        } catch {
            authError = "No se pudo registrar la cuenta."
        }
        isLoading = false
    }

    func logout() {
        keychain.delete(service: tokenService, account: "access")
        keychain.delete(service: tokenService, account: "refresh")
        isAuthenticated = false
        user = nil
    }

    func loadPersistedSession() {
        guard let _ = try? keychain.read(service: tokenService, account: "access") else {
            return
        }
        isAuthenticated = true
    }

    private func persist(tokens: AuthTokens) throws {
        try keychain.save(Data(tokens.accessToken.utf8), service: tokenService, account: "access")
        try keychain.save(Data(tokens.refreshToken.utf8), service: tokenService, account: "refresh")
    }
}
